kaco2pvo/docs/man/man3/kaco2pvo.3
.TH "kaco2pvo" 3 "Tue Jan 20 2015" "kaco2pv" \" -*- nroff -*-
.ad l
.nh
.SH NAME
kaco2pvo \- 
.SH SYNOPSIS
.br
.PP
.SS "Classes"

.in +1c
.ti -1c
.RI "class \fBpowerReading\fP"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "def \fBnum\fP (stringToConvert)"
.br
.ti -1c
.RI "def \fBpost\fP (uri, params)"
.br
.ti -1c
.RI "def \fBpostPVstatus\fP (pvsTimeOfReading, pvsEnergyGen, pvsPowerGen, pvsEnergyUse, pvsPowerUse, pvsTemp, pvsVolts)"
.br
.ti -1c
.RI "def \fBpostPVoutput\fP (pvoDateOfOutput, pvoGenerated, pvoExported, pvoPeakPower, pvoPeakTime, pvoCondition, pvoMinTemp, pvoMaxTemp, pvoImportPeak, pvoImportOffPeak, pvoImportShoulder, pvoImportHighShoulder, pvoConsumption, pvoComment)"
.br
.ti -1c
.RI "def \fBaddReading\fP (newPowerReading)"
.br
.ti -1c
.RI "def \fBprocessReading\fP (readingToProcess)"
.br
.in -1c
.SS "Variables"

.in +1c
.ti -1c
.RI "string \fBLOG_FILENAME\fP = '/var/log/solar/kaco2pv\&.log'"
.br
.ti -1c
.RI "string \fBLOG_READINGS_FILENAME\fP = '/var/log/solar/kaco2pv_readings\&.log'"
.br
.ti -1c
.RI "string \fBformat\fP = '%(asctime)s %(name)-12s %(levelname)-8s %(message)s'"
.br
.ti -1c
.RI "string \fBdatefmt\fP = '%m-%d %H:%M'"
.br
.ti -1c
.RI "\fBfilename\fP = \fBLOG_FILENAME\fP,"
.br
.ti -1c
.RI "string \fBfilemode\fP = 'w'"
.br
.ti -1c
.RI "tuple \fBCONSOLE\fP = logging\&.StreamHandler()"
.br
.ti -1c
.RI "tuple \fBFORMATTER\fP = logging\&.Formatter('%(name)-12s: %(levelname)-8s %(message)s')"
.br
.ti -1c
.RI "tuple \fBLOGGER1\fP = logging\&.getLogger('kaco2pv\&.general')"
.br
.ti -1c
.RI "tuple \fBLOGGER2\fP = logging\&.getLogger('kaco2pv\&.dailyreadings')"
.br
.ti -1c
.RI "tuple \fBLOGGER3\fP = logging\&.getLogger('kaco2pv\&.readings')"
.br
.ti -1c
.RI "tuple \fBLOGGER4\fP = logging\&.getLogger('kaco2pv\&.posting')"
.br
.ti -1c
.RI "tuple \fBLOGGER5\fP = logging\&.getLogger('kaco2pv\&.inverter')"
.br
.ti -1c
.RI "tuple \fBLOGGER6\fP = logging\&.getLogger('kaco2pv\&.raw')"
.br
.ti -1c
.RI "tuple \fBpvdata\fP"
.br
.ti -1c
.RI "tuple \fBformatter\fP = logging\&.Formatter('%(message)s')"
.br
.ti -1c
.RI "string \fBPVS_DEVICE\fP = '/dev/ttyUSB0'"
.br
.ti -1c
.RI "float \fBPVS_MIN_PVP\fP = 7\&.5"
.br
.ti -1c
.RI "float \fBPVS_MAX_PVP\fP = 6000\&.0"
.br
.ti -1c
.RI "string \fBPVO_HOST\fP = 'pvoutput\&.org'"
.br
.ti -1c
.RI "string \fBPVO_STATUS_URI\fP = '/service/r2/addstatus\&.jsp'"
.br
.ti -1c
.RI "string \fBPVO_OUTPUT_URI\fP = '/service/r2/addoutput\&.jsp'"
.br
.ti -1c
.RI "string \fBPVO_KEY\fP = '7ed8a297d387d3887dbd8059c5d8544382a4a12b'"
.br
.ti -1c
.RI "string \fBPVO_SYSTEM_ID\fP = '24657'"
.br
.ti -1c
.RI "int \fBPVO_STATUS_INTERVAL\fP = 5"
.br
.ti -1c
.RI "int \fBSAMPLE_TIME\fP = 10"
.br
.ti -1c
.RI "int \fBPV_DAILY_UPLOAD_TIME_HOUR\fP = 23"
.br
.ti -1c
.RI "int \fBPV_DAILY_UPLOAD_TIME_MIN\fP = 45"
.br
.ti -1c
.RI "tuple \fBTIME_NOW\fP = datetime\&.now()"
.br
.ti -1c
.RI "tuple \fBPV_DAILY_UPLOAD_TIME\fP"
.br
.ti -1c
.RI "\fBLOCAL_ONLY_TESTING\fP = False"
.br
.ti -1c
.RI "float \fB_totalUse\fP = 0\&.0"
.br
.ti -1c
.RI "float \fBtotalGen\fP = 0\&.0"
.br
.ti -1c
.RI "float \fBtotalAmps\fP = 0\&.0"
.br
.ti -1c
.RI "float \fBtotalVolts\fP = 0\&.0"
.br
.ti -1c
.RI "int \fBtotalReadings\fP = 0"
.br
.ti -1c
.RI "\fBlastStatus\fP = datetime\&.min"
.br
.ti -1c
.RI "float \fBdailyUse\fP = 0\&.0"
.br
.ti -1c
.RI "float \fBdailyGen\fP = 0\&.0"
.br
.ti -1c
.RI "float \fBdailyEnergy\fP = 0\&.0"
.br
.ti -1c
.RI "int \fBminTemp\fP = 100"
.br
.ti -1c
.RI "int \fBmaxTemp\fP = -100"
.br
.ti -1c
.RI "int \fBdailyReadings\fP = 0"
.br
.ti -1c
.RI "\fBlastOutput\fP = datetime\&.min"
.br
.ti -1c
.RI "float \fBpeakGen\fP = 0\&.0"
.br
.ti -1c
.RI "tuple \fBpeakTime\fP = datetime\&.now()"
.br
.ti -1c
.RI "int \fBsunriseHour\fP = 9"
.br
.ti -1c
.RI "\fBfullDaysReadings\fP = False"
.br
.in -1c
.SH "Function Documentation"
.PP 
.SS "def kaco2pvo\&.addReading ( newPowerReading)"

.PP
.nf
Add a power reading 
.fi
.PP
 
.PP
Definition at line 310 of file kaco2pvo\&.py\&.
.SS "def kaco2pvo\&.num ( stringToConvert)"

.PP
.nf
returns an int or float from string 
.fi
.PP
 
.PP
Definition at line 179 of file kaco2pvo\&.py\&.
.SS "def kaco2pvo\&.post ( uri,  params)"

.PP
.nf
Performs posting to pvoutput.org 
.fi
.PP
 
.PP
Definition at line 238 of file kaco2pvo\&.py\&.
.SS "def kaco2pvo\&.postPVoutput ( pvoDateOfOutput,  pvoGenerated,  pvoExported,  pvoPeakPower,  pvoPeakTime,  pvoCondition,  pvoMinTemp,  pvoMaxTemp,  pvoImportPeak,  pvoImportOffPeak,  pvoImportShoulder,  pvoImportHighShoulder,  pvoConsumption,  pvoComment)"

.PP
.nf
Create string and pass it to post function 
.fi
.PP
 
.PP
Definition at line 290 of file kaco2pvo\&.py\&.
.SS "def kaco2pvo\&.postPVstatus ( pvsTimeOfReading,  pvsEnergyGen,  pvsPowerGen,  pvsEnergyUse,  pvsPowerUse,  pvsTemp,  pvsVolts)"

.PP
.nf
Create string and pass it to status posting 
.fi
.PP
 
.PP
Definition at line 266 of file kaco2pvo\&.py\&.
.SS "def kaco2pvo\&.processReading ( readingToProcess)"

.PP
.nf
Take string and process into component parts of reading 
.fi
.PP
 
.PP
Definition at line 387 of file kaco2pvo\&.py\&.
.SH "Variable Documentation"
.PP 
.SS "float kaco2pvo\&._totalUse = 0\&.0"

.PP
Definition at line 157 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.CONSOLE = logging\&.StreamHandler()"

.PP
Definition at line 73 of file kaco2pvo\&.py\&.
.SS "float kaco2pvo\&.dailyEnergy = 0\&.0"

.PP
Definition at line 166 of file kaco2pvo\&.py\&.
.SS "float kaco2pvo\&.dailyGen = 0\&.0"

.PP
Definition at line 165 of file kaco2pvo\&.py\&.
.SS "int kaco2pvo\&.dailyReadings = 0"

.PP
Definition at line 169 of file kaco2pvo\&.py\&.
.SS "float kaco2pvo\&.dailyUse = 0\&.0"

.PP
Definition at line 164 of file kaco2pvo\&.py\&.
.SS "string kaco2pvo\&.datefmt = '%m-%d %H:%M'"

.PP
Definition at line 69 of file kaco2pvo\&.py\&.
.SS "string kaco2pvo\&.filemode = 'w'"

.PP
Definition at line 71 of file kaco2pvo\&.py\&.
.SS "kaco2pvo\&.filename = \fBLOG_FILENAME\fP,"

.PP
Definition at line 70 of file kaco2pvo\&.py\&.
.SS "string kaco2pvo\&.format = '%(asctime)s %(name)-12s %(levelname)-8s %(message)s'"

.PP
Definition at line 68 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.FORMATTER = logging\&.Formatter('%(name)-12s: %(levelname)-8s %(message)s')"

.PP
Definition at line 76 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.formatter = logging\&.Formatter('%(message)s')"

.PP
Definition at line 111 of file kaco2pvo\&.py\&.
.SS "kaco2pvo\&.fullDaysReadings = False"

.PP
Definition at line 176 of file kaco2pvo\&.py\&.
.SS "kaco2pvo\&.lastOutput = datetime\&.min"

.PP
Definition at line 170 of file kaco2pvo\&.py\&.
.SS "kaco2pvo\&.lastStatus = datetime\&.min"

.PP
Definition at line 162 of file kaco2pvo\&.py\&.
.SS "kaco2pvo\&.LOCAL_ONLY_TESTING = False"

.PP
Definition at line 155 of file kaco2pvo\&.py\&.
.SS "string kaco2pvo\&.LOG_FILENAME = '/var/log/solar/kaco2pv\&.log'"

.PP
Definition at line 61 of file kaco2pvo\&.py\&.
.SS "string kaco2pvo\&.LOG_READINGS_FILENAME = '/var/log/solar/kaco2pv_readings\&.log'"

.PP
Definition at line 62 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.LOGGER1 = logging\&.getLogger('kaco2pv\&.general')"

.PP
Definition at line 89 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.LOGGER2 = logging\&.getLogger('kaco2pv\&.dailyreadings')"

.PP
Definition at line 91 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.LOGGER3 = logging\&.getLogger('kaco2pv\&.readings')"

.PP
Definition at line 93 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.LOGGER4 = logging\&.getLogger('kaco2pv\&.posting')"

.PP
Definition at line 95 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.LOGGER5 = logging\&.getLogger('kaco2pv\&.inverter')"

.PP
Definition at line 97 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.LOGGER6 = logging\&.getLogger('kaco2pv\&.raw')"

.PP
Definition at line 102 of file kaco2pvo\&.py\&.
.SS "int kaco2pvo\&.maxTemp = -100"

.PP
Definition at line 168 of file kaco2pvo\&.py\&.
.SS "int kaco2pvo\&.minTemp = 100"

.PP
Definition at line 167 of file kaco2pvo\&.py\&.
.SS "float kaco2pvo\&.peakGen = 0\&.0"

.PP
Definition at line 172 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.peakTime = datetime\&.now()"

.PP
Definition at line 173 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.PV_DAILY_UPLOAD_TIME"
\fBInitial value:\fP
.PP
.nf
1 = TIME_NOW\&.replace(hour=PV_DAILY_UPLOAD_TIME_HOUR,
2                                         minute=PV_DAILY_UPLOAD_TIME_MIN,
3                                         second=0, microsecond=0)
.fi
.PP
Definition at line 150 of file kaco2pvo\&.py\&.
.SS "int kaco2pvo\&.PV_DAILY_UPLOAD_TIME_HOUR = 23"

.PP
Definition at line 145 of file kaco2pvo\&.py\&.
.SS "int kaco2pvo\&.PV_DAILY_UPLOAD_TIME_MIN = 45"

.PP
Definition at line 146 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.pvdata"
\fBInitial value:\fP
.PP
.nf
1 = logging\&.handlers\&.TimedRotatingFileHandler(LOG_READINGS_FILENAME,
2                                                    when='midnight',
3                                                    interval=1,
4                                                    backupCount=14,
5                                                    encoding=None,
6                                                    delay=False,
7                                                    utc=False)
.fi
.PP
Definition at line 103 of file kaco2pvo\&.py\&.
.SS "string kaco2pvo\&.PVO_HOST = 'pvoutput\&.org'"

.PP
Definition at line 132 of file kaco2pvo\&.py\&.
.SS "string kaco2pvo\&.PVO_KEY = '7ed8a297d387d3887dbd8059c5d8544382a4a12b'"

.PP
Definition at line 137 of file kaco2pvo\&.py\&.
.SS "string kaco2pvo\&.PVO_OUTPUT_URI = '/service/r2/addoutput\&.jsp'"

.PP
Definition at line 134 of file kaco2pvo\&.py\&.
.SS "int kaco2pvo\&.PVO_STATUS_INTERVAL = 5"

.PP
Definition at line 141 of file kaco2pvo\&.py\&.
.SS "string kaco2pvo\&.PVO_STATUS_URI = '/service/r2/addstatus\&.jsp'"

.PP
Definition at line 133 of file kaco2pvo\&.py\&.
.SS "string kaco2pvo\&.PVO_SYSTEM_ID = '24657'"

.PP
Definition at line 139 of file kaco2pvo\&.py\&.
.SS "string kaco2pvo\&.PVS_DEVICE = '/dev/ttyUSB0'"

.PP
Definition at line 126 of file kaco2pvo\&.py\&.
.SS "float kaco2pvo\&.PVS_MAX_PVP = 6000\&.0"

.PP
Definition at line 130 of file kaco2pvo\&.py\&.
.SS "float kaco2pvo\&.PVS_MIN_PVP = 7\&.5"

.PP
Definition at line 128 of file kaco2pvo\&.py\&.
.SS "int kaco2pvo\&.SAMPLE_TIME = 10"

.PP
Definition at line 143 of file kaco2pvo\&.py\&.
.SS "int kaco2pvo\&.sunriseHour = 9"

.PP
Definition at line 175 of file kaco2pvo\&.py\&.
.SS "tuple kaco2pvo\&.TIME_NOW = datetime\&.now()"

.PP
Definition at line 149 of file kaco2pvo\&.py\&.
.SS "float kaco2pvo\&.totalAmps = 0\&.0"

.PP
Definition at line 159 of file kaco2pvo\&.py\&.
.SS "float kaco2pvo\&.totalGen = 0\&.0"

.PP
Definition at line 158 of file kaco2pvo\&.py\&.
.SS "int kaco2pvo\&.totalReadings = 0"

.PP
Definition at line 161 of file kaco2pvo\&.py\&.
.SS "float kaco2pvo\&.totalVolts = 0\&.0"

.PP
Definition at line 160 of file kaco2pvo\&.py\&.
.SH "Author"
.PP 
Generated automatically by Doxygen for kaco2pv from the source code\&.


kaco2pvo.py
#!/usr/bin/env python
""" @package docstring
Documentation for this module
    
Process to take serial data from Kaco inverter and upload
to pvoutput.org
# Upload from Kaco Powador serial output to pvoutput.org live feed
#
#
# To be run on RaspberryPi Python3 (IDLE3) using Prolific USB/Serial cable
# To download required serial library from terminal prompt:
#
# sudo apt-get install python3-serial
#
# Supplied with limited tested for others to tailor to their own needs
#
# This software in any form is covered by the following Open Source BSD license
#
# Copyright 2013-2014, Ian Hutt
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
# PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER
# OR CONTRIBUTORS BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITE
# TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
# OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
# STRICT LIABILITY, OR TORT (INCLUDING
# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
"""


from datetime import datetime
import serial
import sys

#Import requirements for posting
import http.client
import urllib, urllib.parse

# Import requirements for logging
import logging
import logging.handlers

LOG_FILENAME = '/var/log/solar/kaco2pv.log'
LOG_READINGS_FILENAME = '/var/log/solar/kaco2pv_readings.log'


# Setup logging
# set up logging to file
logging.basicConfig(level=logging.DEBUG,
                    format='%(asctime)s %(name)-12s %(levelname)-8s %(message)s',
                    datefmt='%m-%d %H:%M',
                    filename=LOG_FILENAME,
                    filemode='w')
# define a Handler which writes INFO messages or higher to the sys.stderr
CONSOLE = logging.StreamHandler()
CONSOLE.setLevel(logging.DEBUG) #was info
# set a format which is simpler for console use
FORMATTER = logging.Formatter('%(name)-12s: %(levelname)-8s %(message)s')
# tell the handler to use this format
CONSOLE.setFormatter(FORMATTER)
# add the handler to the root logger
logging.getLogger('').addHandler(CONSOLE)

# Now, we can log to the root logger, or any other logger. First the root...
# logging.info('Jackdaws love my big sphinx of quartz.')

# Now, define a couple of other loggers which might represent areas in your
# application:

# general logging
LOGGER1 = logging.getLogger('kaco2pv.general')
# logging info for daily output
LOGGER2 = logging.getLogger('kaco2pv.dailyreadings')
# logging info for regular output
LOGGER3 = logging.getLogger('kaco2pv.readings')
# logging info for posting
LOGGER4 = logging.getLogger('kaco2pv.posting')
# logging info for inverter
LOGGER5 = logging.getLogger('kaco2pv.inverter')


# Now logging specifically of the data received from the inverter

LOGGER6 = logging.getLogger('kaco2pv.raw')
pvdata = logging.handlers.TimedRotatingFileHandler(LOG_READINGS_FILENAME,
                                                   when='midnight',
                                                   interval=1,
                                                   backupCount=14,
                                                   encoding=None,
                                                   delay=False,
                                                   utc=False)
pvdata.setLevel(logging.INFO)
formatter = logging.Formatter('%(message)s')
pvdata.setFormatter(formatter)
LOGGER6.addHandler(pvdata)

# Example log usage
# LOGGER1.debug('Quick zephyrs blow, vexing daft Jim.')
# LOGGER1.info('How quickly daft jumping zebras vex.')
# LOGGER2.warning('Jail zesty vixen who grabbed pay from quack.')
# LOGGER2.error('The five boxing wizards jump quickly.')

LOGGER1.info("pvs2pvo - (c) Ian Hutt 2014")
LOGGER1.info("modified by Chris Peck for Kaco Inverters")
LOGGER1.info("Startup")

# May need changing
PVS_DEVICE = "/dev/ttyUSB0"
# PV generation values (W) less than this will be set to 0
PVS_MIN_PVP = 7.5
# PV generation values greater than this will be set to this
PVS_MAX_PVP = 6000.0

PVO_HOST = "pvoutput.org"
PVO_STATUS_URI = "/service/r2/addstatus.jsp"
PVO_OUTPUT_URI = "/service/r2/addoutput.jsp"


PVO_KEY = "7ed8a297d387d3887dbd8059c5d8544382a4a12b"
# Your PVoutput system ID here
PVO_SYSTEM_ID = "24657"
# How often in minutes to update PVoutput
PVO_STATUS_INTERVAL = 5
# Time in hours of updates from Kaco unit (normally 10 seconds)
SAMPLE_TIME = 10/3600

PV_DAILY_UPLOAD_TIME_HOUR = 23
PV_DAILY_UPLOAD_TIME_MIN = 45


TIME_NOW = datetime.now()
PV_DAILY_UPLOAD_TIME = TIME_NOW.replace(hour=PV_DAILY_UPLOAD_TIME_HOUR,
                                        minute=PV_DAILY_UPLOAD_TIME_MIN,
                                        second=0, microsecond=0)

# True for local only, False turns on pvoutput upload
LOCAL_ONLY_TESTING = False

_totalUse = 0.0
totalGen = 0.0
totalAmps = 0.0
totalVolts = 0.0
totalReadings = 0
lastStatus = datetime.min

dailyUse = 0.0
dailyGen = 0.0
dailyEnergy = 0.0
minTemp = 100
maxTemp = -100
dailyReadings = 0
lastOutput = datetime.min
# set last daily summary output to epoch ie never
peakGen = 0.0
peakTime = datetime.now()
# latest time after which it is decided that there aren't a full days readings
sunriseHour = 9
fullDaysReadings = False


def num(stringToConvert):
    """ returns an int or float from string """
    try:
        return int(stringToConvert)
    except ValueError:
        return float(stringToConvert)


class powerReading:
    """ class containing power reading data """
    _timeOfReading = datetime.min
    _dailyRunTime = 0
    _operatingState = 0
    _generatorVoltage = 0.0
    _generatorCurrent = 0.0
    _generatorPower = 0.0
    _lineVoltage = 0.0
    _lineCurrentFeedIn = 0.0
    _powerFeedIn = 0.0
    _unitTemperature = 0.0

    def timeOfReading(self):
        """ returns time of reading """
        return self._timeOfReading

    def generatedPower(self):
        """ returns generated power """
        return self._generatorPower

    def temperature(self):
        """ returns temperature """
        return self._unitTemperature

    def generatedVoltage(self):
        """ returns generated voltage """
        return self._generatorVoltage

    def generatedCurrent(self):
        """ returns generated current """
        return self._generatorCurrent

    def __init__(self, timeOfReading, dailyRunTime, operatingState,
                 generatorVoltage,
                 generatorCurrent,
                 generatorPower, lineVoltage, lineCurrentFeedIn,
                 powerFeedIn, unitTemperature):
        """ class initialiser """
        self._timeOfReading = timeOfReading
        self._dailyRunTime = dailyRunTime
        self._operatingState = operatingState
        self._generatorVoltage = generatorVoltage
        self._generatorCurrent = generatorCurrent
        self._generatorPower = generatorPower
        self._lineVoltage = lineVoltage
        self._lineCurrentFeedIn = lineCurrentFeedIn
        self._powerFeedIn = powerFeedIn
        self._unitTemperature = unitTemperature


def post(uri, params):
    """ Performs posting to pvoutput.org """
    if not LOCAL_ONLY_TESTING:
        try:
            LOGGER4.debug("Posting results - " + str(params))
            headers = {'X-Pvoutput-Apikey': PVO_KEY,
                       'X-Pvoutput-SystemId': PVO_SYSTEM_ID,
                       "Accept": "text/plain",
                       "Content-type": "application/x-www-form-urlencoded"}
            conn = http.client.HTTPConnection(PVO_HOST)
#               conn.set_debuglevel(2) # debug purposes only
            conn.request("POST", uri, urllib.parse.urlencode(params), headers)
            response = conn.getresponse()
            LOGGER4.debug("Status" + str(response.status) + "   Reason:" +
                          str(response.reason) + "-" + str(response.read()))
            sys.stdout.flush()
            conn.close()
            return response.status == 200
        except Exception as e:
            LOGGER4.error("Exception posting results\n" + str(e))
            sys.stdout.flush()
            return False
    else:
        LOGGER4.debug("In testing mode - NOT posting results")
        return True


def postPVstatus(pvsTimeOfReading, pvsEnergyGen,
                 pvsPowerGen, pvsEnergyUse, pvsPowerUse, pvsTemp, pvsVolts):
    """ Create string and pass it to status posting """
    LOGGER4.debug("In postPVstatus")
    LOGGER4.debug("Constructing params")
    params = {'d': pvsTimeOfReading.strftime('%Y%m%d'),
              't': pvsTimeOfReading.strftime('%H:%M'),
              'v1': pvsEnergyGen,
              'v2': pvsPowerGen,
              'v3': pvsEnergyUse,  # for later use if required
              'v4': pvsPowerUse,
              'v5': pvsTemp,
              'v6': pvsVolts,
              'c1': 0,
              'n': 0}

    LOGGER4.debug("Params:" + str(params))
    # POST the data
    return post(PVO_STATUS_URI, params)


def postPVoutput(pvoDateOfOutput, pvoGenerated, pvoExported, pvoPeakPower,
                 pvoPeakTime,
                 pvoCondition, pvoMinTemp, pvoMaxTemp,
                 pvoImportPeak, pvoImportOffPeak, pvoImportShoulder,
                 pvoImportHighShoulder, pvoConsumption, pvoComment):
    """ Create string and pass it to post function """
    params = {'d': pvoDateOfOutput.strftime('%Y%m%d'),
              'g': pvoGenerated,
              'e': pvoExported,
              'pp': pvoPeakPower,
              'pt': pvoPeakTime.strftime('%H:%M'),
              'tm': pvoMinTemp,
              'tx': pvoMaxTemp,
              'ip': pvoImportPeak,
              'io': pvoImportOffPeak,
              'is': pvoImportShoulder,
              'ih': pvoImportHighShoulder,
              'c': pvoConsumption,
              'cm': 'EOD upload.'
                    + pvoComment}
    LOGGER4.debug("Params:" + str(params))
    return post(PVO_OUTPUT_URI, params)


def addReading(newPowerReading):
    """ Add a power reading """
    gen = newPowerReading.generatedPower()
    volts = newPowerReading.generatedVoltage()
    temperature = newPowerReading.temperature()
    amps = newPowerReading.generatedCurrent()
    timeNow = datetime.now()

    comment = ""
    global totalGen, totalReadings, lastStatus, totalAmps, totalVolts
    totalGen += gen
    totalAmps += amps
    totalVolts += volts
    totalReadings += 1
    global dailyEnergy, dailyGen, dailyReadings
    global lastOutput, peakGen, peakTime, fullDaysReadings, maxTemp, minTemp
    dailyGen += gen
    dailyEnergy += (gen*SAMPLE_TIME)
    dailyReadings += 1
    LOGGER3.info("Gen:" + str(gen) + "W " + str(dailyEnergy) + "Wh "
                 + str(volts) + "V " + str(amps) + "A")
    if gen > peakGen:
        peakGen = int(gen)
        peakTime = timeNow
    if temperature > maxTemp:
        maxTemp = temperature
    if temperature < minTemp:
        minTemp = temperature
    if (timeNow.minute % PVO_STATUS_INTERVAL == 0) & (lastStatus.minute != timeNow.minute):
        # You may wish to modify the simplistic averaging for something
        # more sophisticated like weighted average or kalman
        avgGen = float(totalGen / totalReadings)
        avgVoltage = float(totalVolts / totalReadings)
        avgCurrent = float(totalAmps / totalReadings)
        LOGGER3.info("Time to output status. avgGen:" +
                     str(avgGen) + "W " + str(avgVoltage) + "V " + str(avgCurrent) + "A")
        if postPVstatus(timeNow, dailyEnergy,
                        avgGen, 0, 0, temperature, avgVoltage):
            lastStatus = timeNow
            totalGen = 0.0
            totalAmps = 0.0
            totalVolts = 0.0
            totalReadings = 0
            LOGGER3.info("Status update sucessfully sent to PVoutput")
        else:
            lastStatus = timeNow
            LOGGER3.error("Failed to post status to pvoutput")
        sys.stdout.flush()
    LOGGER2.debug("Debugging daily summary")
    LOGGER2.debug("timeNow.hour = " + str(timeNow.hour))
    LOGGER2.debug("timeNow.minute = " + str(timeNow.minute))
    LOGGER2.debug("timeNow.day = " + str(timeNow.day))
    LOGGER2.debug("lastOutput.day = " + str(lastOutput.day))
    # Check whether its time to upload daily results
    # The time has to be after the PV_DAILY_UPLOAD_TIME
    # The lastOutput either has to have been never(epoch) or not today 
    if ((timeNow > PV_DAILY_UPLOAD_TIME) and ( (timeNow.day != lastOutput.day) or (lastOutput == datetime.min))):
        daysGen = int((dailyGen / dailyReadings) * 24.0)
        LOGGER2.info("Time to output EOD")
        LOGGER2.info("DaysGen:" + str(daysGen) + "W")
        if not fullDaysReadings:
            LOGGER2.info("Incomplete days readings.")
            comment = "Incomplete readings for days"
        if postPVoutput(timeNow, daysGen, 0, peakGen, peakTime, 'Not Sure',
                        minTemp, maxTemp, 0, 0, 0, 0, 0, comment):
            lastOutput = timeNow
            dailyGen = 0
            dailyReadings = 0
            peakGen = 0
            peakTime = timeNow
            minTemp = 100
            maxTemp = -100
            dailyEnergy = 0
            LOGGER2.info("EOD Output sucessfully sent to PVoutput")
        else:
            LOGGER2.error("Failed to post daily output to pvoutput")
        fullDaysReadings = True
    sys.stdout.flush()


def processReading(readingToProcess):
    """ Take string and process into component parts of reading """
    LOGGER6.info(readingToProcess)
    LOGGER5.debug("Reading data" + readingToProcess)
    myReadings = readingToProcess.split()
    placeHolder = myReadings[0]
    LOGGER5.debug("placeHolder = " + placeHolder)
    dailyRunTime = myReadings[1]
    LOGGER5.debug("dailyRunTime = " + str(dailyRunTime))
    operatingState = myReadings[2]
    LOGGER5.debug("operatingState = " + str(operatingState))
    generatorVoltage = num(myReadings[3])
    LOGGER5.debug("generatorVoltage = " + str(generatorVoltage))
    generatorCurrent = num(myReadings[4])
    LOGGER5.debug("generatorCurrent = " + str(generatorCurrent))
    generatorPower = num(myReadings[5])
    LOGGER5.debug("generatorPower = " + str(generatorPower))
    lineVoltage = num(myReadings[6])
    LOGGER5.debug("lineVoltage = " + str(lineVoltage))
    lineCurrentFeedIn = num(myReadings[7])
    LOGGER5.debug("lineCurrentFeedIn = " + str(lineCurrentFeedIn))
    powerFeedIn = num(myReadings[8])
    LOGGER5.debug("powerFeedIn = " + str(powerFeedIn))
    unitTemperature = num(myReadings[9])
    LOGGER5.debug("unitTemperature = " + str(unitTemperature))
    LOGGER5.debug("Processing data")
    if generatorPower < PVS_MIN_PVP:
        generatorPower = 0
    elif generatorPower > PVS_MAX_PVP:
        generatorPower = PVS_MAX_PVP

    newReading = powerReading(datetime.utcnow(), dailyRunTime, operatingState,
                              generatorVoltage, generatorCurrent,
                              generatorPower, lineVoltage, lineCurrentFeedIn,
                              powerFeedIn, unitTemperature)
    addReading(newReading)

try:
    LOGGER5.info("Opening PV inverter serial port on " + str(PVS_DEVICE))
    COM_PORT = serial.Serial(PVS_DEVICE, baudrate=9600, bytesize=8,
                             parity='N', stopbits=1, xonxoff=0, timeout=5.0)
    COM_PORT.open()

    if COM_PORT.isOpen():
        COM_PORT.flush()
        LOGGER5.info("PV inverter connection opened")
    else:
        LOGGER5.error("Unable to open connection to PV inverter")

    LOGGER5.info("Processing PV inverter data")

    START_TIME = datetime.now()
    if START_TIME.hour > sunriseHour:
        fullDaysReadings = False
        LOGGER5.info("Not a full days readings")
    else:
        fullDaysReadings = True
        LOGGER5.info("A full days readings")

    sys.stdout.flush()
    myBuffer = ''
    while COM_PORT.isOpen():
        try:
            myBuffer += COM_PORT.read(max(1, COM_PORT.inWaiting())).decode('utf-8', "replace")
        except Exception as ex:
            LOGGER5.error("Exception reading from PV inverter. Terminating\n" + str(ex))
            LOGGER5.error("Check that another process is not using device")
            COM_PORT.close()
            sys.stdout.flush()
        if '\r' in myBuffer:
            readings = myBuffer.split('\r')
            myBuffer = readings.pop()
            for newReading in readings:
                processReading(newReading)

    COM_PORT.close()
except Exception as e:
    LOGGER1.error("Exception processing PV inverter data. Terminating")
    LOGGER1.error(str(e))
    LOGGER1.error(str(sys.exc_info()[0]))
    sys.stdout.flush()
    raise


LOGGER1.info("Bye")
sys.stdout.flush()
